### PR TITLE
ROX-19047: Add more specific error messages to sensor integration test

### PR DESCRIPTION
**Pull request by fredrb**
## Description

Show an error message that no alert results were found if fake central doesn't have alerts for requested deployment. Right now, this integration test will fail with the assertion message, which is misleading. 

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] ~~Unit test and regression tests added~~
- [ ] ~~Evaluated and added CHANGELOG entry if required~~
- [ ] ~~Determined and documented upgrade steps~~
- [ ] ~~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

- CI

### CODE REVIEW

**Comment by vikin91** [comment]
[Recent failure](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/stackrox_stackrox/7405/pull-ci-stackrox-stackrox-master-gke-sensor-integration-tests/1691162041785520128) was not about the test case being first.
```
=== RUN   Test_NetworkPolicy/Test_Deployment_NetpolViolations/Both_policies_applied:_should_have_no_violations
    helper.go:636: alerts found for deployment nginx-deployment: no violations for deployment
=== RUN   Test_NetworkPolicy/Test_Deployment_NetpolViolations/Ingress_applied:_egress_violation
=== RUN   Test_NetworkPolicy/Test_Deployment_NetpolViolations/Egress_applied:_ingress_violation
--- FAIL: Test_NetworkPolicy (46.71s)
    --- FAIL: Test_NetworkPolicy/Test_Deployment_NetpolViolations (46.34s)
        --- PASS: Test_NetworkPolicy/Test_Deployment_NetpolViolations/No_policies_applied:_should_have_two_violations (12.04s)
        --- FAIL: Test_NetworkPolicy/Test_Deployment_NetpolViolations/Both_policies_applied:_should_have_no_violations (11.54s)
        --- PASS: Test_NetworkPolicy/Test_Deployment_NetpolViolations/Ingress_applied:_egress_violation (11.37s)
        --- PASS: Test_NetworkPolicy/Test_Deployment_NetpolViolations/Egress_applied:_ingress_violation (11.39s)
FAIL
```

File: sensor/tests/helper/helper.go [hunk 1]
```diff
@@ -618,6 +621,24 @@ func (c *TestContext) LastViolationStateWithTimeout(t *testing.T, name string, a
...
+	ticker := time.NewTicker(defaultTicker)
+	for {
+		select {
+		case <-timer.C:
+			return
+		case <-ticker.C:
+			messages := c.GetFakeCentral().GetAllMessages()
+			alerts := GetAllAlertsForDeploymentName(messages, name)
+			if len(alerts) > 0 {
+				t.Fatalf("alerts found for deployment %s: %s", name, message)
```
Comment (vikin91) [inline sensor/tests/helper/helper.go:636]: Can we print the alerts here for debugging?

**Comment by vikin91** [comment]
It looks like the alerts are still there after we apply the policies:

```
=== RUN   Test_NetworkPolicy/Test_Deployment_NetpolViolations/Both_policies_applied:_should_have_no_violations
    helper.go:636: alerts found for deployment nginx-deployment ([event:<id:"4f76e198-3cb7-43e2-b5c2-648cce5cf1f4" action:SYNC_RESOURCE sensor_hash:6472742790193243251 alert_results:<deployment_id:"4f76e198-3cb7-43e2-b5c2-648cce5cf1f4" alerts:<id:"25ac1c68-f945-477b-822c-965efeeb5b6f" policy:<id:"32d770b9-c6ba-4398-b48a-0c3e807644ed" name:"Deployments should have at least one egress Network Policy" description:"Alerts if deployments are missing an egress Network Policy" categories:"Security Best Practices" lifecycle_stages:DEPLOY exclusions:<name:"Don't alert on kube-system namespace" deployment:<scope:<namespace:"kube-system" > > > exclusions:<name:"Don't alert on openshift-kube-apiserver namespace" deployment:<scope:<namespace:"openshift-kube-apiserver" > > > exclusions:<name:"Don't alert on openshift-kube-scheduler namespace" deployment:<scope:<namespace:"openshift-kube-scheduler" > > > exclusions:<name:"Don't alert on openshift-kube-controller-manager namespace" deployment:<scope:<namespace:"openshift-kube-controller-manager" > > > exclusions:<name:"Don't alert on openshift-sdn namespace" deployment:<scope:<namespace:"openshift-sdn" > > > exclusions:<name:"Don't alert on openshift-network-operator namespace" deployment:<scope:<namespace:"openshift-network-operator" > > >
```

Shall we add some waiting? ❄️❄️❄️

File: sensor/tests/resource/networkpolicy/networkpolicy_test.go [hunk 2]
```diff
@@ -99,7 +99,11 @@ func (s *NetworkPolicySuite) Test_Deployment_NetpolViolations() {
 			s.testContext.RunTest(t,
 				helper.WithResources(resourcesToApply),
 				helper.WithTestCase(func(t *testing.T, tc *helper.TestContext, objects map[string]k8s.Object) {
-					tc.LastViolationState(t, "nginx-deployment", checkViolations(testCase.violationsExpected), name)
+					if len(testCase.violationsExpected) == 0 {
+						tc.NoViolationForDeployment(t, "nginx-deployment", "no violations for deployment")
+					} else {
+						tc.LastViolationState(t, "nginx-deployment", checkViolations(testCase.violationsExpected), name)
+					}
```
Comment (vikin91) [inline sensor/tests/resource/networkpolicy/networkpolicy_test.go:106]: It looks like the existing violations need a moment to disappear.
Shall we maybe retry that block for a while until the conditions are met?

Comment (fredrb) [reply sensor/tests/resource/networkpolicy/networkpolicy_test.go]: Changed the check a little to fix this. Let's see in the next run.

**Comment by vikin91** [comment]
I see we still have somewhere the `testing.go:1471: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test}`. Let's fix that in a follow-up.

I see some different tests started failing 🤔

**Comment by fredrb** [comment]
> I see we still have somewhere the testing.go:1471: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test}. Let's fix that in a follow-up.

These are from services tests, and are just surfacing the same problem that I fixed in the imagescan tests. The tests were expecting that _some_ alert result should be sent, and that it doesn't contain the violation we are looking for. When in reality we need to assert that _no_ alert result object is ever sent. 

This will probably always fail now due to [this change](https://github.com/stackrox/stackrox/pull/7405/files#diff-e55206bf349f1b0af43261904577fd58cb1264b8c25747b14bf90cc65239704fR610-R612) I made.

I think it should be fixed in my latest commit. I ran the test locally and it always failed before the commit and now it's passing. Let's see what CI says.

**Review by vikin91 — APPROVED** [review]
LGTM! 

I triggered one more run for gke integration tests. Let's make sure this is fixed now

**Comment by vikin91** [comment]
It passed :) Nice! I think we can merge it

**Pull request merged**